File: sdks/python/src/echo_agent_sdk/prompt_cache_layout_values.py
from __future__ import annotations

from collections.abc import Mapping, Sequence
from dataclasses import dataclass, field
from typing import Any

from .segment_range_values import SegmentRange


@dataclass(frozen=True, slots=True)
class SegmentRanges:
    """Message-cache segment indexes without owning cache state."""

    system: SegmentRange = field(default_factory=SegmentRange)
    canonical: SegmentRange = field(default_factory=SegmentRange)
    history: SegmentRange = field(default_factory=SegmentRange)
    runtime_context: SegmentRange = field(default_factory=SegmentRange)


@dataclass(frozen=True, slots=True)
class PromptCacheLayout:
    """Read-only prompt layout projection; cache placement stays in Rust."""

    system: tuple[Mapping[str, Any], ...]
    canonical: tuple[Mapping[str, Any], ...]
    history: tuple[Mapping[str, Any], ...]
    runtime_context: tuple[Mapping[str, Any], ...]
    tools: tuple[Mapping[str, Any], ...]

    @classmethod
    def from_messages(
        cls,
        messages: Sequence[Mapping[str, Any]],
        tools: Sequence[Mapping[str, Any]],
    ) -> PromptCacheLayout:
        if not isinstance(messages, Sequence) or isinstance(messages, (str, bytes)):
            raise TypeError("messages must be a sequence")
        if not isinstance(tools, Sequence) or isinstance(tools, (str, bytes)):
            raise TypeError("tools must be a sequence")
        message_values = _snapshot_values(messages, "messages")
        tool_values = _snapshot_values(tools, "tools")

        system_end = next(
            (
                index
                for index, message in enumerate(message_values)
                if message.get("role") != "system"
            ),
            len(message_values),
        )
        canonical_start = next(
            (
                index
                for index, message in enumerate(message_values[:system_end])
                if (_message_text(message) or "").find("Canonical context") >= 0
            ),
            system_end,
        )

        runtime_start = len(message_values)
        for index in range(len(message_values) - 1, -1, -1):
            text = _message_text(message_values[index])
            if text is not None and text.lstrip().startswith("[runtime_context:"):
                runtime_start = index
                while runtime_start > system_end:
                    previous = _message_text(message_values[runtime_start - 1])
                    if previous is None or not previous.lstrip().startswith(
                        "[runtime_context:"
                    ):
                        break
                    runtime_start -= 1
                break

        return cls(
            tuple(message_values[: min(canonical_start, system_end)]),
            tuple(message_values[canonical_start:system_end])
            if canonical_start < system_end
            else (),
            tuple(message_values[system_end:runtime_start]),
            tuple(message_values[runtime_start:]),
            tuple(tool_values),
        )

    def segment_ranges(self) -> SegmentRanges:
        system = len(self.system)
        canonical = len(self.canonical)
        history = len(self.history)
        runtime_context = len(self.runtime_context)
        canonical_end = system + canonical
        history_end = canonical_end + history
        return SegmentRanges(
            system=SegmentRange(0, system),
            canonical=SegmentRange(system, canonical_end),
            history=SegmentRange(canonical_end, history_end),
            runtime_context=SegmentRange(history_end, history_end + runtime_context),
        )


def _snapshot_values(
    values: Sequence[Mapping[str, Any]], name: str
) -> list[dict[str, Any]]:
    result: list[dict[str, Any]] = []
    for value in values:
        if not isinstance(value, Mapping):
            raise TypeError(f"{name} entries must be objects")
        result.append(dict(value))
    return result


def _message_text(message: Mapping[str, Any]) -> str | None:
    content = message.get("content")
    if isinstance(content, str):
        return content
    if isinstance(content, Mapping) and content.get("kind") == "string":
        value = content.get("value")
        return value if isinstance(value, str) else None
    return None

File: sdks/typescript/src/prompt_cache_layout_values.ts
import type { LlmMessage, LlmToolDefinition, WireValue } from "./types.js";
import { SegmentRange } from "./segment_range_values.js";

/** Message-cache segment indexes without owning cache state. */
export class SegmentRanges {
  public constructor(
    public readonly system: SegmentRange = new SegmentRange(),
    public readonly canonical: SegmentRange = new SegmentRange(),
    public readonly history: SegmentRange = new SegmentRange(),
    public readonly runtimeContext: SegmentRange = new SegmentRange(),
  ) {
    Object.freeze(this);
  }
}

/** Read-only prompt layout projection; cache storage and provider placement stay in Rust. */
export class PromptCacheLayout {
  public readonly system: readonly LlmMessage[];
  public readonly canonical: readonly LlmMessage[];
  public readonly history: readonly LlmMessage[];
  public readonly runtimeContext: readonly LlmMessage[];
  public readonly tools: readonly LlmToolDefinition[];

  private constructor(
    system: readonly LlmMessage[],
    canonical: readonly LlmMessage[],
    history: readonly LlmMessage[],
    runtimeContext: readonly LlmMessage[],
    tools: readonly LlmToolDefinition[],
  ) {
    this.system = Object.freeze([...system]);
    this.canonical = Object.freeze([...canonical]);
    this.history = Object.freeze([...history]);
    this.runtimeContext = Object.freeze([...runtimeContext]);
    this.tools = Object.freeze([...tools]);
    Object.freeze(this);
  }

  public static fromMessages(
    messages: readonly LlmMessage[],
    tools: readonly LlmToolDefinition[],
  ): PromptCacheLayout {
    if (!Array.isArray(messages) || !Array.isArray(tools)) {
      throw new TypeError("messages and tools must be arrays");
    }
    const systemEnd = messages.findIndex((message) => message.role !== "system");
    const sysEnd = systemEnd < 0 ? messages.length : systemEnd;
    const canonicalIndex = messages
      .slice(0, sysEnd)
      .findIndex((message) => messageText(message)?.includes("Canonical context") ?? false);
    const canonicalStart = canonicalIndex < 0 ? sysEnd : canonicalIndex;

    let runtimeStart = messages.length;
    for (let index = messages.length - 1; index >= 0; index -= 1) {
      const candidate = messages[index];
      if (candidate && (messageText(candidate)?.trimStart().startsWith("[runtime_context:") ?? false)) {
        runtimeStart = index;
        while (runtimeStart > sysEnd) {
          const previous = messages[runtimeStart - 1];
          if (!previous || !(messageText(previous)?.trimStart().startsWith("[runtime_context:") ?? false)) break;
          runtimeStart -= 1;
        }
        break;
      }
    }

    return new PromptCacheLayout(
      messages.slice(0, Math.min(canonicalStart, sysEnd)),
      canonicalStart < sysEnd ? messages.slice(canonicalStart, sysEnd) : [],
      messages.slice(sysEnd, runtimeStart),
      messages.slice(runtimeStart),
      tools,
    );
  }

  public segmentRanges(): SegmentRanges {
    const system = BigInt(this.system.length);
    const canonical = BigInt(this.canonical.length);
    const history = BigInt(this.history.length);
    const runtimeContext = BigInt(this.runtimeContext.length);
    return new SegmentRanges(
      new SegmentRange(0n, system),
      new SegmentRange(system, system + canonical),
      new SegmentRange(system + canonical, system + canonical + history),
      new SegmentRange(system + canonical + history, system + canonical + history + runtimeContext),
    );
  }
}

function messageText(message: LlmMessage): string | null {
  const content: unknown = message.content;
  if (typeof content === "string") return content;
  if (isWireString(content)) return content.value;
  return null;
}

function isWireString(value: unknown): value is Extract<WireValue, { readonly kind: "string" }> {
  return typeof value === "object"
    && value !== null
    && !Array.isArray(value)
    && (value as { readonly kind?: unknown }).kind === "string"
    && typeof (value as { readonly value?: unknown }).value === "string";
}
